src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -566,23 +566,15 @@ public static AbstractMemorySegmentImpl ofBuffer(Buffer bb) {
     }
 
     private static int getScaleFactor(Buffer buffer) {
-        if (buffer instanceof ByteBuffer) {
-            return 0;
-        } else if (buffer instanceof CharBuffer) {
-            return 1;
-        } else if (buffer instanceof ShortBuffer) {
-            return 1;
-        } else if (buffer instanceof IntBuffer) {
-            return 2;
-        } else if (buffer instanceof FloatBuffer) {
-            return 2;
-        } else if (buffer instanceof LongBuffer) {
-            return 3;
-        } else if (buffer instanceof DoubleBuffer) {
-            return 3;
-        } else {
-            throw new AssertionError("Cannot get here");
-        }
+        return switch (buffer) {
```

Comment:
sorry, this is, I think, buggy. Note that you can have an IntBuffer wrapped around a byte[], in which case the calculation will misfire. I think the factory using this method should just call BaseAndScale::of (on the heap array) and then get the scale from there.

Reply:
We should probably add a test to detect regressions such as these

Reply:
Perhaps we can do that under a separate issue as it is a broader effort than just modernizing the code?

Reply:
Please, ignore this comment. I looked again at the code, and the scale factor is used to scale the buffer's position. Of course, the position needs to be scaled depending on the buffer kind (e.g. an int buffer has a scaling factor of 4, a short buffer has a scaling factor of 2). The underlying array type doesn't matter when scaling the buffer position: the position is a logical index expressed in elements, so only the buffer type matters (which is what the code does).